### ISSUE TITLE
[Not a Bug / Installation] Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED

### ISSUE THREAD

**Issue by invadetl** [mono/mono#12406]
I am trying to test mono / msbuild on RHEL 8 Beta.

Mono 5.18.0.225 has been built and installed successfully.

MSBuild 16.0 has been installed but running "make" returns a number of CERTIFICATE_VERIFY_FAILED errors.

On this server, libssl (1.1.1) is installed in /usr/lib64/:
lrwxrwxrwx. 1 root root     15 Oct 15 10:17 /usr//lib64/libssl.so.1.1 -> libssl.so.1.1.1
-rwxr-xr-x. 1 root root 629160 Oct 15 10:18 /usr//lib64/libssl.so.1.1.1

I have also tried installing the 32-bit version into /usr/lib/ plus the compatible version, 1.0, but these made no difference.

I'm not sure if the build process is using OpenSSL from Red Hat or Boring SSL from the Mono installation.

I ran mozroots, but then any execution of:
certmgr -ssl {URL}
returns:
*** WARNING: Certificate signature is INVALID ***
for every single URL.

Have I missed a step from the installation process?

Please let me know if you need any more information.

## Steps to Reproduce

1. Download msbuild from https://download.mono-project.com/sources/msbuild/msbuild-16.0+xamarinxplat.2018.09.26.17.53.tar.xz
2. Extract tarball.
3. Run make.

### Current Behavior

Various certificate failures, eg.

/root/msbuild-16.0/artifacts/mono-msbuild/NuGet.targets(114,5): error : Unable to load the service index for source https://dotnet.myget.org/F/roslyn/api/v3/index.json. [/root/msbuild-16.0/src/Build/Microsoft.Build.csproj]
/root/msbuild-16.0/artifacts/mono-msbuild/NuGet.targets(114,5): error :   An error occurred while sending the request [/root/msbuild-16.0/src/Build/Microsoft.Build.csproj]
/root/msbuild-16.0/artifacts/mono-msbuild/NuGet.targets(114,5): error :   Error: TrustFailure (Authentication failed, see inner exception.) [/root/msbuild-16.0/src/Build/Microsoft.Build.csproj]
/root/msbuild-16.0/artifacts/mono-msbuild/NuGet.targets(114,5): error :   Authentication failed, see inner exception. [/root/msbuild-16.0/src/Build/Microsoft.Build.csproj]
/root/msbuild-16.0/artifacts/mono-msbuild/NuGet.targets(114,5): error :   Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED [/root/msbuild-16.0/src/Build/Microsoft.Build.csproj]
/root/msbuild-16.0/artifacts/mono-msbuild/NuGet.targets(114,5): error :     at /root/mono-5.18.0.225/external/boringssl/ssl/handshake_client.c:1132 [/root/msbuild-16.0/src/Build/Microsoft.Build.csproj]

### Expected Behavior

Software builds successfully.

## On which platforms did you notice this

[ ] macOS
[x ] Linux
[ ] Windows

RHEL 8 Beta
Mono 5.18.0.225

**Comment by baulig** [comment]
This is an installation issue that we need to document better as it is confusing and hard to understand to new people.

Mono does not use the installed OpenSsl, but instead ships a built-in bundled copy of boringssl.  It also uses its own root certificate store, so the `mozroots` tool is not going to help.

You need to run the `cert-sync` tool after installing Mono the first time.

FIXME: this is a documentation bug - we should document this somewhere on the web site.

**Comment by invadeint** [comment]
Many thanks for the response.

I can confirm that running the following has resolved the problem:
cert-sync /etc/ssl/certs/ca-bundle.crt

**Comment by embix** [comment]
The fix obviously only works if you have a ca-bundle. BLFS systemd-9.0 for example does provide it ([unless you upgrade from an old version](http://www.linuxfromscratch.org/blfs/view/stable-systemd/postlfs/make-ca.html)), there still is an [open issue in systemd for providing a trusted-CA bundle on boot](https://github.com/systemd/systemd/issues/13242).

In that case you'd have to generate a current ca-bundle first and import that:

```
p11-kit extract --format=pem-bundle --filter=ca-anchors --comment --purpose server-auth ./tls-ca-bundle.$(date -d "today" +"%Y%m%d%H%M").pem
cert-sync ./tls-ca-bundle.201910022003.pem
```

The flaw with the bundle/external database is, that they tend to get out of sync eventually, unless you automate syncing (i.e. via cronjob). The mess is obvious if you look at the [Fedora Wiki page for SharedSystemCertificates](https://fedoraproject.org/wiki/Features/SharedSystemCertificates).

**Comment by bizmonger** [comment]
I believe I'm running into the same issue.
Unfortunately, I'm not well-versed in security certificates and how to properly install them.

Is there any simple tutorials on how to install whatever is necessary for this error to be resolved?

**Comment by schultetwin1** [comment]
@baulig - I am running into this issue intermittently on [Azure Pipeline's Ubuntu 18.04 image](https://github.com/actions/virtual-environments/blob/master/images/linux/Ubuntu1804-README.md). I have attempted to run

```bash
sudo update-ca-certificates && sudo cert-sync /etc/ssl/certs/ca-certificates.crt
```

before using mono to run `nuget` but this doesn't seem to fix the error. I believe this may be an issue with the generated VM and have opened an issue actions/virtual-environments#1270 with GitHub to track this down, but need help at what to look for.

Is there any tracing or logging that will help us figure out which certificate mono is using to attempt to authenticate

**Comment by schultetwin1** [comment]
@baulig - I added the following traces to mono when running nuget.exe

```
mono --trace=N:Mono.Net.Security,N:Mono.Btls
```

And was able to get a trace on both the failure case (bad case) and a successful case (good case). Note, they are (unfortunately) very verbose. The good case file is also trimmed just cut down on the size of the file.

[mono_tracing_bad_case.txt](https://github.com/mono/mono/files/5011366/mono_tracing_bad_case.txt)
[mono_tracing_good_case.txt](https://github.com/mono/mono/files/5011370/mono_tracing_good_case.txt)

I also now have what seems to be a workaround for this issue with mono + nuget.exe. Adding the flag `-DisableParallelProcessing` to nuget.exe, which prevents multiple threads from downloading in parallel.

Note: Without `-DisableParallelProcessing` this issue repro'd about 6 times in 96 attempts (on Azure Pipeline's Ubuntu 18.04 image). With `-DisableParallelProcessing` this issue never repro'd in 200 attempts (on Azure Pipeline's Ubuntu 18.04 image).

**Comment by darkdragon-001** [comment]
This is currently blocking me in flathub/com.github.PintaProject.Pinta#14 for compiling [https://github.com/GtkSharp/GtkSharp](GtkSharp) for Flatpak.

Running `cert-sync` without `--user` doesn't work because of a read-only file system. Paths need to be prefixed with `/app`. The following script doesn't work either:

    cert-sync --user /etc/ssl/certs/ca-certificates.crt
    mkdir -p /app/usr/share/.mono/certs
    mv $HOME/.config/.mono/certs/* /app/usr/share/.mono/certs/

**Comment by pracplayopen** [comment]
Have experienced this problem before and cert-sync usually fixes it.

However today experiencing this even after cert-sync has imported all machine's certs.

I'm getting CERTIFICATE_VERIFY_FAILS for certs whose issuing CA is visible INSIDE of certmgr.

```
$ mono5 /mnt/sda1/mono/ver/5.18.0/usr/local/lib/mono/4.5/csharp.exe  -e 'new System.Net.WebClient ().DownloadString
 ("https://www.nuget.org")'
System.Net.WebException: Error: TrustFailure (Authentication failed, see inner exception.) ---> System.Security.Authentication.AuthenticationException: Authentication failed, see inner exception. ---> Mono.Btls.MonoBtlsException: Ssl error:1000007d:SSL routines:OPENSSL_internal:CERTIFICATE_VERIFY_FAILED
  at /mnt/sda1/mono/mono/external/boringssl/ssl/handshake_client.c:1132
  at Mono.Btls.MonoBtlsContext.ProcessHandshake () [0x00038] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at Mono.Net.Security.MobileAuthenticatedStream.ProcessHandshake (Mono.Net.Security.AsyncOperationStatus status, System.Boolean renegotiate) [0x000a1] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at (wrapper remoting-invoke-with-check) Mono.Net.Security.MobileAuthenticatedStream.ProcessHandshake(Mono.Net.Security.AsyncOperationStatus,bool)
  at Mono.Net.Security.AsyncHandshakeRequest.Run (Mono.Net.Security.AsyncOperationStatus status) [0x00006] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at Mono.Net.Security.AsyncProtocolRequest.ProcessOperation (System.Threading.CancellationToken cancellationToken) [0x000ff] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at Mono.Net.Security.AsyncProtocolRequest.StartOperation (System.Threading.CancellationToken cancellationToken) [0x0008b] in <64cd34170db1432ea65220b2e4102b7f>:0 
   --- End of inner exception stack trace ---
  at Mono.Net.Security.MobileAuthenticatedStream.ProcessAuthentication (System.Boolean runSynchronously, Mono.Net.Security.MonoSslAuthenticationOptions options, System.Threading.CancellationToken cancellationToken) [0x00252] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at Mono.Net.Security.MonoTlsStream.CreateStream (System.Net.WebConnectionTunnel tunnel, System.Threading.CancellationToken cancellationToken) [0x00126] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebConnection.CreateStream (System.Net.WebOperation operation, System.Boolean reused, System.Threading.CancellationToken cancellationToken) [0x001ba] in <64cd34170db1432ea65220b2e4102b7f>:0 
   --- End of inner exception stack trace ---
  at System.Net.WebConnection.CreateStream (System.Net.WebOperation operation, System.Boolean reused, System.Threading.CancellationToken cancellationToken) [0x0021a] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebConnection.InitConnection (System.Net.WebOperation operation, System.Threading.CancellationToken cancellationToken) [0x00141] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebOperation.Run () [0x0009a] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebCompletionSource`1[T].WaitForCompletion () [0x00094] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.HttpWebRequest.RunWithTimeoutWorker[T] (System.Threading.Tasks.Task`1[TResult] workerTask, System.Int32 timeout, System.Action abort, System.Func`1[TResult] aborted, System.Threading.CancellationTokenSource cts) [0x000f8] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.HttpWebRequest.GetResponse () [0x00016] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebClient.GetWebResponse (System.Net.WebRequest request) [0x00000] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebClient.DownloadBits (System.Net.WebRequest request, System.IO.Stream writeStream) [0x000e6] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebClient.DownloadDataInternal (System.Uri address, System.Net.WebRequest& request) [0x00061] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebClient.DownloadString (System.Uri address) [0x00011] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at System.Net.WebClient.DownloadString (System.String address) [0x00008] in <64cd34170db1432ea65220b2e4102b7f>:0 
  at <InteractiveExpressionClass>.Host (System.Object& $retval) [0x00006] in <d32e0f6d688549ed8de5418a35b5e3e2>:0 
  at Mono.CSharp.Evaluator.Evaluate (System.String input, System.Object& result, System.Boolean& result_set) [0x00038] in <56ef5380aa204410840c18429d38b3a4>:0 
  at Mono.CSharpShell.Evaluate (System.String input) [0x00000] in <3751d28b63294cf594692f5f376a2548>:0 

$ $ mono5 /mnt/sda1/mono/ver/5.18.0/u
sr/local/lib/mono/4.5/certmgr.exe  -list -c Trust|grep -A4 -B3 "DigiCert Global Root G2"

Self-signed X.509 v3 Certificate
  Serial Number: E5FA091DB16428BBA0A911A7E6F13A03
  Issuer Name:   C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Global Root G2
  Subject Name:  C=US, O=DigiCert Inc, OU=www.digicert.com, CN=DigiCert Global Root G2
  Valid From:    8/1/2013 12:00:00 PM
  Valid Until:   1/15/2038 12:00:00 PM
  Unique Hash:   4B4EB4B074298B828B5C003095A10B4523FB951C0C88348B09C53E5BABA408A3

```

This is the issuer of nuget's certificate and the serial # in certmgr exactly matches what is on the cert in a browser.

I can also find the cert installed in ~/.config/.mono/certs.   

So I'm struggling to collect information on why doesn't it verify.  Any suggestions?   

I've tried running with MONO_DEBUG enabled but it doesn't show any specific error that is causing validation to fail.   

I've tried removing all the certs and reimporting/syncing them to mono, same deal.  

I've tried installing certs manually with certmgr -ssl, but I also get an authentication error when I do this.    Also I don't think this is the issue since I can see the root certs in certmgr.

I may try to compare traces from two machines running same version of mono and see what is different (as previous user did, but didn't specify what he saw when reviewing them).

Any ideas appreciated.

**Comment by marxin** [comment]
> Any ideas appreciated.

I can confirm that I also face the same problem and `cert-sync` hasn't helped me. I created an upstream openSUSE bug for it:
https://bugzilla.opensuse.org/show_bug.cgi?id=1191220

Have you came to something that fixed that?

Thank you.

**Comment by marxin** [comment]
> I can confirm that I also face the same problem and `cert-sync` hasn't helped me. I created an upstream openSUSE bug for it: https://bugzilla.opensuse.org/show_bug.cgi?id=1191220

So I investigated the change is related to `DST Root CA X3 Expiration (September 2021)`:
https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/

It used to work yesterday.

**Comment by marxin** [comment]
> So I investigated the change is related to `DST Root CA X3 Expiration (September 2021)`: https://letsencrypt.org/docs/dst-root-ca-x3-expiration-september-2021/
> 
> It used to work yesterday.

Got it. It's related to BoringSSL issue that can be fixed with:
https://github.com/electron/electron/pull/31213

Note the upstream library is not fixed yet and apparently mono contains it's own copy of BoringSSL library.

**Comment by marxin** [comment]
New issue I created for it #21233.

**Comment by HolimaX** [comment]
Dupe of https://github.com/mono/mono/issues/21438 ?

**Comment by estahn** [comment]
If you come here due to Sonarr and Rdarr not working. SSH into your Synology and run:
```
/volume1/@appstore/mono/bin/cert-sync /etc/ssl/certs/ca-certificates.crt
```

This will sync the built-in certs with the system certs. Restart your Synology.

**Comment by chillichicken** [comment]
Make that 

```
sudo /volume1/@appstore/mono/bin/cert-sync /etc/ssl/certs/ca-certificates.crt
```
